### PR TITLE
Implement kube addon manager

### LINKED ISSUE #665: Ensure we handle deleting of no longer configured kubernetesa resources
**What you expected to happen**:

If Tarmaks puppet code removes certain manifests it should ensure that those are getting removed as well

**How to reproduce it (as minimally and precisely as possible)**:

E.g. disable and enable the deployment of Prometheus resources

**Anything else we need to know?**:

This might be solved by use kube-addon-manager

### PR DESCRIPTION
**Pull request by JoshVanL**
Uses kube addon-manager to manage the life cycle of resources created by tarmak.

fixes #665

```release-note
Deploy Kubernetes resources using kube-addon-manager 
```

### CODE REVIEW

**Comment by simonswine** [comment]
There seems to be a race between the masters (so somehow the election of kube-addon-managers and/or cluster-manager is not stable.

```
tarmak cluster kubectl get pods --all-namespaces -w
kube-system   heapster-687467cd4c-wmkbb   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-wmkbb   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-wmkbb   0/2   ContainerCreating   0     0s
kube-system   heapster-687467cd4c-wmkbb   2/2   Running   0     3s
kube-system   heapster-789c695b5-ztw8b   2/2   Terminating   0     52s
kube-system   heapster-789c695b5-ztw8b   0/2   Terminating   0     53s
kube-system   heapster-789c695b5-ztw8b   0/2   Terminating   0     54s
kube-system   heapster-789c695b5-ztw8b   0/2   Terminating   0     54s
kube-system   heapster-789c695b5-kqh5w   0/2   Pending   0     0s
kube-system   heapster-789c695b5-kqh5w   0/2   Pending   0     0s
kube-system   heapster-789c695b5-kqh5w   0/2   ContainerCreating   0     0s
kube-system   heapster-789c695b5-kqh5w   2/2   Running   0     3s
kube-system   heapster-687467cd4c-wmkbb   2/2   Terminating   0     13s
kube-system   heapster-687467cd4c-wmkbb   0/2   Terminating   0     15s
kube-system   heapster-687467cd4c-wmkbb   0/2   Terminating   0     16s
kube-system   heapster-687467cd4c-wmkbb   0/2   Terminating   0     16s
^[kube-system   heapster-687467cd4c-qphn2   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-qphn2   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-qphn2   0/2   ContainerCreating   0     0s
kube-system   heapster-687467cd4c-qphn2   2/2   Running   0     4s
kube-system   heapster-789c695b5-kqh5w   2/2   Terminating   0     53s
kube-system   heapster-789c695b5-kqh5w   0/2   Terminating   0     54s
kube-system   heapster-789c695b5-vdf9p   0/2   Pending   0     0s
kube-system   heapster-789c695b5-vdf9p   0/2   Pending   0     0s
kube-system   heapster-789c695b5-vdf9p   0/2   ContainerCreating   0     0s
kube-system   heapster-789c695b5-kqh5w   0/2   Terminating   0     61s
kube-system   heapster-789c695b5-kqh5w   0/2   Terminating   0     61s
kube-system   heapster-789c695b5-vdf9p   2/2   Running   0     4s
kube-system   heapster-687467cd4c-qphn2   2/2   Terminating   0     15s
kube-system   heapster-687467cd4c-qphn2   0/2   Terminating   0     16s
kube-system   heapster-687467cd4c-qphn2   0/2   Terminating   0     28s
kube-system   heapster-687467cd4c-qphn2   0/2   Terminating   0     28s
kube-system   heapster-687467cd4c-jhsw7   0/2   Pending   0     1s
kube-system   heapster-687467cd4c-jhsw7   0/2   Pending   0     1s
kube-system   heapster-687467cd4c-jhsw7   0/2   ContainerCreating   0     1s
kube-system   heapster-687467cd4c-jhsw7   2/2   Running   0     4s
kube-system   heapster-789c695b5-vdf9p   2/2   Terminating   0     53s
kube-system   heapster-789c695b5-vdf9p   0/2   Terminating   0     55s
kube-system   heapster-789c695b5-vdf9p   0/2   Terminating   0     56s
kube-system   heapster-789c695b5-vdf9p   0/2   Terminating   0     56s
kube-system   heapster-789c695b5-tp48k   0/2   Pending   0     0s
kube-system   heapster-789c695b5-tp48k   0/2   Pending   0     0s
kube-system   heapster-789c695b5-tp48k   0/2   ContainerCreating   0     0s
kube-system   heapster-789c695b5-tp48k   2/2   Running   0     3s
kube-system   heapster-687467cd4c-jhsw7   2/2   Terminating   0     17s
kube-system   heapster-687467cd4c-jhsw7   0/2   Terminating   0     17s
kube-system   heapster-687467cd4c-jhsw7   0/2   Terminating   0     18s
kube-system   heapster-687467cd4c-jhsw7   0/2   Terminating   0     18s
kube-system   heapster-687467cd4c-744gp   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-744gp   0/2   Pending   0     0s
kube-system   heapster-687467cd4c-744gp   0/2   ContainerCreating   0     0s
kube-system   heapster-687467cd4c-744gp   2/2   Running   0     4s
kube-system   heapster-789c695b5-tp48k   2/2   Terminating   0     50s
```

This is from a cluster with three master nodes:

```
NAME                                          STATUS   ROLES    AGE   VERSION   INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                KERNEL-VERSION               CONTAINER-RUNTIME   ROLE

ip-10-99-108-76.eu-west-1.compute.internal    Ready    master   17m   v1.13.1   10.99.108.76    <none>        CentOS Linux 7 (Core)   4.20.0-1.el7.elrepo.x86_64   docker://1.13.1     master
ip-10-99-47-197.eu-west-1.compute.internal    Ready    master   18m   v1.13.1   10.99.47.197    <none>        CentOS Linux 7 (Core)   4.20.0-1.el7.elrepo.x86_64   docker://1.13.1     master
ip-10-99-64-87.eu-west-1.compute.internal     Ready    master   17m   v1.13.1   10.99.64.87     <none>        CentOS Linux 7 (Core)   4.20.0-1.el7.elrepo.x86_64   docker://1.13.1     master
```

diff between the two RSs is:
```
61,62c61,62
<             cpu: 100m
<             memory: 150Mi
---
>             cpu: 108m
>             memory: 214Mi
64,65c64,65
<             cpu: 100m
<             memory: 150Mi
---
>             cpu: 108m
>             memory: 214Mi
```

/assign @JoshVanL 
/unassign

File: puppet/modules/calico/templates/node-rbac.yaml.erb [hunk 1]
```diff
@@ -25,6 +27,8 @@ apiVersion: rbac.authorization.k8s.io/v1beta1
 kind: ClusterRoleBinding
 metadata:
   name: calico-node
+  labels:
+    addonmanager.kubernetes.io/mode: EnsureExists
```
Comment (simonswine) [inline puppet/modules/calico/templates/node-rbac.yaml.erb]: Just a bit below: ServiceAccount is missing

File: puppet/modules/kubernetes/manifests/apply.pp [hunk 2]
```diff
@@ -47,34 +40,37 @@
...
+    $protocol = 'http'
   }
-  -> service{ "${service_name}.service":
-    ensure  => 'running',
-    enable  => true,
-    require => [
-      Service[$service_apiserver],
-    ]
+
+  $command = "/bin/bash -c \"while true; do if [[ \$(curl -k -w '%{http_code}' -s -o /dev/null ${protocol}://localhost:${server_port}/healthz) == 200 ]]; then break; fi; done; kubectl apply -f '${apply_file}' || rm -f '${apply_file})'\""
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply puppet/modules/kubernetes/manifests/apply.pp]: Yep, good idea. I did at one point during testing but must have removed it...

**Comment by simonswine** [comment]
Good work

/approve
/lgtm

**Pull request merged**